README.md

```diff
@@ -33,6 +33,285 @@ of this, and doing so can be easier said than done, especially if your algorithm
+    pos.push(uv::Vec3x8::new(pos_x, pos_y, pos_z));
+    vel.push(uv::Vec3x8::new(vel_x, vel_y, vel_z));
+    acc.push(uv::Vec3x8::new(acc_x, acc_y, acc_z));
+}
+```
+
+Note that `100 / 8` in maths terms would be `12.5`, but we can't conveniently have a half-sized `Vec3x8`.
+
+There are various ways to handle these 'remainder' vectors. You could fall back to scalar code, or progressively fall back to narrower wide types, such as `Vec3x4`, or you can just consider whether the cost of calculating a few additional vectors that you won't use is worth adding complexity to your code.
+
+### Ray-Sphere Intersection
+
+Scalar code that operates on a single value at a time needs some restructuring to take advantage of SIMD and the 4-/8-wide data types.
+
+Below is an example of scalar ray-sphere instersection code using `Vec3` for points and vectors:
+
+```rust
+fn ray_sphere_intersect(
+    ray_o: uv::Vec3,
+    ray_d: uv::Vec3,
+    sphere_o: uv::Vec3,
+    sphere_r_sq: f32,
+) -> f32 {
+    let oc = ray_o - sphere_o;
+    let b = oc.dot(ray_d);
+    let c = oc.mag_sq() - sphere_r_sq;
+    let descrim = b * b - c;
+
+    if descrim > 0.0 {
+        let desc_sqrt = descrim.sqrt();
+
+        let t1 = -b - desc_sqrt;
+        if t1 > 0.0 {
+            t1
+        } else {
+            let t2 = -b + desc_sqrt;
+            if t2 > 0.0 {
+                t2
+            } else {
+                f32::MAX
+            }
+        }
+    } else {
+        f32::MAX
+    }
+}
+```
+
+This porting guide will not discuss the details of the algorithm, but will focus on how to convert the code to apply parallel SIMD operations on wide data types.
+
+The first thing to do is to convert the parameter types from scalar `Vec3` to wide `Vec3x8` and `f32x8`:
+
+```rust
+fn ray_sphere_intersect_x8(
+    ray_o: uv::Vec3x8,
+    ray_d: uv::Vec3x8,
+    sphere_o: uv::Vec3x8,
+    sphere_r_sq: uv::f32x8,
+) -> uv::f32x8 {
+```
+
+Each call to the function will process 8 ray-sphere intersections in parallel. The first four lines of the function remain the same:
+
+```rust
+    let oc = ray_o - sphere_o;
+    let b = oc.dot(ray_d);
+    let c = oc.mag_sq() - sphere_r_sq;
+    let descrim = b * b - c;
+```
+
+Despite this code being the same, the calculations for 8 rays and spheres will be carried out at the same time!
+
+The next line of the scalar code tests the value of `descrim` to see if it is greater than `0.0`. When operating on 8 values at a time, the code cannot branch along two separate paths because the value of `descrim` for each of the 8 values may cause branching to different sets of operations. To support this we would need to convert back to scalar code and then we lose all the performance benefits of parallel processing.
+
+So, how do we convert this? It turns out, the trick is to do the calculations for both branches for all 8 lanes, and then filter the results with masks that select the correct values from the possibilities at the end.
```

Comment:
I think it's important to note here that this decision-making process has some nuance. For example, depending on the frequency of divergence, it might be worth it to check whether all lanes are on one side or the other of the check and then branch accordingly. Those cases are relatively rare though

Reply:
I tried to comment on that a bit further down but I’ll add something in. Thanks for the suggestion about checking the common case being a strategy in such a situation.

Reply:
I made an attempt at this. I feel like it's a bit too verbose, but maybe it's useful. What do you think?